tfstate/remote_s3.go

```diff
@@ -6,24 +6,26 @@ import (
 	"path"
 
 	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/s3"
 	"github.com/aws/aws-sdk-go/service/s3/s3manager"
 )
 
 func readS3State(config map[string]interface{}, ws string) (io.ReadCloser, error) {
-	region, bucket, key := *strp(config["region"]), *strp(config["bucket"]), *strp(config["key"])
+	role, region, bucket, key := *strpe(config["role_arn"]), *strpe(config["region"]), *strpe(config["bucket"]), *strpe(config["key"])
 	if ws != defaultWorkspace {
 		if prefix := strp(config["workspace_key_prefix"]); prefix != nil {
 			key = path.Join(*prefix, ws, key)
 		} else {
 			key = path.Join(defaultWorkspeceKeyPrefix, ws, key)
 		}
 	}
-	return readS3(region, bucket, key)
+	return readS3(role, region, bucket, key)
 }
 
-func readS3(region, bucket, key string) (io.ReadCloser, error) {
+func readS3(role, region, bucket, key string) (io.ReadCloser, error) {
```

Comment:
S3 State Storage requires bucket and key only.
https://www.terraform.io/docs/language/settings/backends/s3.html#s3-state-storage

The other parameters are optional. So readS3() should have an interface like `readS3(bucket, key string, opt s3Option)` I think.

```
type s3Option struct {
   region string
   role string
}

func readS3(bucket, key string, opt s3Option) (io.ReadCloser, error) {
```

This interface can accept extra elements in `config` in the future without argument changes.

Reply:
OK!
Change to the s3Option structure and push again.

Reply:
@fujiwara 
Does it match your implementation plan?
https://github.com/fujiwara/tfstate-lookup/pull/22/commits/2a0b7f1bd3199450d7e96cbe1b2d5f7ff8745374